resolwe/flow/managers/base.py

```diff
@@ -74,26 +168,365 @@ def discover_engines(self):
+        override = self.state.settings_override or {}
+        override.update(immediates)
+        self.settings_actual = self._marshal_settings()
+        self.settings_actual.update(override)
+
+        if cmd == WorkerProtocol.COMMUNICATE:
+            try:
+                self._communicate(**message.content[WorkerProtocol.COMMUNICATE_EXTRA])
+            finally:
+                new_sema = self.state.sync_semaphore.add(-1)  # clear communicate() claim on the semaphore
+                logger.debug(__("Manager changed sync_semaphore DOWN to {} after _communicate().", new_sema))
+
+        elif cmd == WorkerProtocol.FINISH:
+            data_id = message.content[WorkerProtocol.DATA_ID]
+            if not getattr(settings, 'FLOW_MANAGER_TEST_KEEP_DATA', False):
+                shutil.rmtree(self._get_per_data_dir('RUNTIME_DIR', data_id))
+
+            if message.content[WorkerProtocol.FINISH_SPAWNED]:
+                new_sema = self.state.sync_semaphore.add(1)
+                logger.debug(__("Manager changed sync_semaphore UP to {} in spawn handler.", new_sema))
+                try:
+                    self._communicate(**message.content[WorkerProtocol.FINISH_COMMUNICATE_EXTRA])
+                finally:
+                    new_sema = self.state.sync_semaphore.add(-1)  # clear communicate() claim on the semaphore
+                    logger.debug(__("Manager changed sync_semaphore DOWN to {} after _communicate().", new_sema))
+
+            new_count = self.state.executor_count.add(-1)
+            new_sema = self.state.sync_semaphore.add(-1)  # clear execution claim on the semaphore
+            logger.debug(__("Manager changed sync_semaphore DOWN to {} after executor finish.", new_sema))
+
+            if int(self.state.sync_execution) == 1 and new_count == 0:
+                # wake someone waiting for us up
+                Channel(state.MANAGER_SYNC_PIPE).send({})
+
+        else:
+            logger.error(__("Ignoring unknown manager control command `{}`.", cmd))
+
+    def synchronized(self, force_enter=False):
+        """Enter a synchronization block for the calling context.
+
+        :param force_enter: If ``True``, enter the block even if another
+            one is already active. Use this at your own risk.
+        """
+        return self._SynchronizationManager(self.state.key_prefix, force_enter, state.MANAGER_SYNC_PIPE)
+
+    def execution_barrier(self, force_enter=False):
+        """Wait for executors to finish.
+
+        At least one must finish after this point to avoid a deadlock.
+
+        :param force_enter: If ``True``, create the barrier even if
+            another synchronization block is already active. Use this at
+            your own risk.
+        """
+        with self.synchronized(force_enter=force_enter):
+            self.communicate(verbosity=0)
+
+    def communicate(self, run_sync=False, verbosity=1, save_settings=True):
+        """Resolve task dependencies and run the task.
```

Comment:
Can you make the description more accurate?

It leads one to thing that task dependency resolution and execution happens here but it in fact, only `communicate` command is send to manager's channel.

Reply:
I left this summary as it was before the refactor.